go/cli/mcap/cmd/filter.go

```diff
@@ -0,0 +1,326 @@
+package cmd
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"math"
+	"os"
+	"regexp"
+
+	"github.com/foxglove/mcap/go/cli/mcap/utils"
+	"github.com/foxglove/mcap/go/mcap"
+	"github.com/spf13/cobra"
+)
+
+var (
+	filterOutput             string
+	filterIncludeTopics      []string
+	filterExcludeTopics      []string
+	filterStart              uint64
+	filterEnd                uint64
+	filterIncludeMetadata    bool
+	filterIncludeAttachments bool
+	filterOutputCompression  string
+)
+
+type filterOpts struct {
+	includeTopics      []regexp.Regexp
+	excludeTopics      []regexp.Regexp
+	start              uint64
+	end                uint64
+	includeMetadata    bool
+	includeAttachments bool
+	compressionFormat  mcap.CompressionFormat
+}
+
+func buildFilterOptions() (*filterOpts, error) {
+	opts := &filterOpts{
+		includeMetadata:    filterIncludeMetadata,
+		includeAttachments: filterIncludeAttachments,
+	}
+	opts.start = filterStart * 1e9
+	if filterEnd == 0 {
+		opts.end = math.MaxUint64
+	} else {
+		opts.end = filterEnd * 1e9
+	}
+	if len(filterIncludeTopics) > 0 && len(filterExcludeTopics) > 0 {
+		return nil, errors.New("can only use one of --include-topic-regex and --exclude-topic-regex")
+	}
+	if filterEnd < filterStart {
+		return nil, errors.New("invalid time range query, end-time is before start-time")
+	}
+	opts.compressionFormat = mcap.CompressionNone
+	switch filterOutputCompression {
+	case "zstd":
+		opts.compressionFormat = mcap.CompressionZSTD
+	case "lz4":
+		opts.compressionFormat = mcap.CompressionLZ4
+	case "none":
+		opts.compressionFormat = mcap.CompressionNone
+	default:
+		return nil, fmt.Errorf("unrecognized compression format '%s': valid options are 'lz4', 'zstd', or 'none'", filterOutputCompression)
+	}
+
+	includeTopics, err := compileMatchers(filterIncludeTopics)
+	if err != nil {
+		return nil, err
+	}
+	opts.includeTopics = includeTopics
+
+	excludeTopics, err := compileMatchers(filterExcludeTopics)
+	if err != nil {
+		return nil, err
+	}
+	opts.excludeTopics = excludeTopics
+	return opts, nil
+}
+
+// filterCmd represents the filter command
+var filterCmd = &cobra.Command{
+	Use:   "filter",
+	Short: "copy some filtered MCAP data to a new file",
+	Long: `This subcommand filters an MCAP by topic and time range to a new file.
+When multiple regexes are used, topics that match any regex are included (or excluded).
+
+usage example:
+  mcap filter in.mcap -o out.mcap -y /diagnostics -y /tf -y /camera_(front|back)`,
+	Run: func(cmd *cobra.Command, args []string) {
+		filterOptions, err := buildFilterOptions()
+		if err != nil {
+			die("configuration error: %s", err)
+		}
+		var reader io.Reader
+		if len(args) == 0 {
+			stat, err := os.Stdin.Stat()
+			if err != nil {
+				die("failed to check stdin state: %s", err)
+			}
+			if stat.Mode()&os.ModeCharDevice == 0 {
+				reader = os.Stdin
+			} else {
+				die("supply a file to read")
```

Comment:
maybe we can point them at `mcap filter -h` or say `mcap filter [file]` here somehow